types.go

```diff
@@ -67,5 +71,18 @@ type DeviceDesc struct {
 }
 
 type InstanceManager interface {
-	DeviceFromId(name string) (DeviceDesc, error)
+	DeviceFromId(name string, user UserInfo) (DeviceDesc, error)
+}
+
+type AuthHTTPHandler func(http.ResponseWriter, *http.Request, UserInfo) error
+type HTTPHandler func(http.ResponseWriter, *http.Request) error
+
+type AccountManager interface {
+	// Returns the received http handler wrapped in another that extracts user
+	// information from the request and passes it to to the original handler as
+	// the last parameter.
+	// The wrapper will only pass the request to the inner handler if a user is
+	// authenticated, otherwise it may choose to return an error or respond with
+	// an HTTP redirect to the login page.
+	Authorize(fn AuthHTTPHandler) HTTPHandler
```

Comment:
Authenticate instead? Authorize is more related to permissions over resources.

Reply:
I think Authorize makes more sense here:
We are not trying to identify in a secure way who the user really is (authentication), GAE takes care of that for us, but instead we're deciding if a user should be allowed to execute this request (authorization). It just so happens that all (authenticated) users are allowed to do all operations for now. In the future we could extend this function to take the handler and a role that is allowed to make the request so it would check that the user has that role.

Reply:
All it's doing now is still only authentication, and the API doesn't look like it's really set up to make authorization convenient. As soon as the authorization gets more complex than "user X can access all endpoints and user Y cannot", this would either have to reimplement the request routing table to discover which per-request logic it has to run, or the request handler has to do its own authorization, again leaving this as purely authentication (whether relying on app engine information or otherwise).

Reply:
The one thing this api is missing to be obviously authorization is an extra parameter of type UserRole (one of ADMIN, VIEWER, EDITOR, for example). The role parameter would be passed along with the http handler and the implementation of this method would check that the current user is the given role. I didn't add the extra parameter because we don't have a clear idea of what the roles will be yet.

Even without the extra parameter I still think this is authorization. My understanding of the subject can be summarized as:
"authentication is the process of verifying who someone is , whereas authorization is the process of verifying what things a user has access to"
It's impossible to do authorization without doing authentication first, so it's very easy to mix the two, however this method is "deciding whether a request should be allowed to proceed based on who the user is" which, in my opinion, matches the concept of authorization better than it does with authentication.

That being said, I could be wrong. Given that both of you seem to agree I'll rename it, we can continue the debate on another time and rename back if we all agree to do so.